text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1722 @@
+
+```rust
+impl<Scope> PromiseTransmutableFrom<Scope, (NeglectAlignment, NeglectValidity)> for Foo
+{
+    ...
+}
+
+impl<Scope> PromiseTransmutableFrom<Scope, (NeglectValidity, NeglectAlignment)> for Foo
+{
+    ...
+}
+```
+Writing *both* impls (as we do above) is logically nonsense, but is nonetheless supported by Rust's coherence rules.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+
+## Rationale: `TransmuteFrom`/`TransmuteInto`
+
+### Why support arbitrary transmutation?
+Some [prior art][prior-art], especially in the crate ecosystem, provides an API that [only supports transmutations involving particular types](#Source-and-Destination-Types-Supported) (e.g., from/into bytes). As we discuss in the [prior art][prior-art] section, we believe that the inflexibility of such approaches make them a poor basis of a language proposal. In particular, these restrictive approaches don't leave room to grow: supporting additional transmutations requires additional traits.
+
+The API advocated by this proposal is unopinionated about what transmutations users might wish to do, and what transmutations the compiler is able to reason about. The implementation of this RFC may be initially very simple (and perhaps support no more than the restrictive approaches allow for), but then subsequently grow in sophistication—*without* necessitating public API changes.
+
+### Why *two* traits?
+If `TransmuteInto` is implemented in terms of `TransmuteFrom`, why provide it at all? We do so for consistency with libcore's [`From`/`Into`](https://doc.rust-lang.org/stable/rust-by-example/conversion/from_into.html) traits, and because directionality conveys intent: `TransmuteFrom` connotes *conversion*, whereas `TransmuteInto` connotes initialization. We believe that the supporting code examples of this RFC demonstrate the explanatory benefits of providing *both* traits.
+
+## Rationale: Transmutation Options
+
+### Granularity
+Although the focus of our API is statically-correct, infalible transmutations, the ability to opt-out of particular static checks is essential for building safer *fallible* mechanisms, such as alignment-fallible [reference casting][ext-ref-casting], or validity-fallible transmutations (e.g., `bool` to `u8`).
+
+### Representation
+Although transmutations options exist at a type-level, they're represented as type-level tuples, whose familiar syntax is identical to value-level tuples. An empty tuple seems like the natural choice for encoding *don't neglect anything*.
+
+We could not identify any advantages to representing options with const-generics. There is no clear syntactic advantage: tuples remain the most natural way to encode ad-hoc products of items. The comparative lack of default values for const-generic parameters poses an ergonomic *disadvantage*.
```

Comment:
The lack of defaults in const generics is interesting here.  If `= []` worked, would that be preferred?  Intuitively here I was expecting one `#[non_exhaustive]` enum, not a sealed trait with N non-instantiable types and a bunch of wants-variadic-generics impls of the trait.  I guess the normal case here is passing one thing, so passing `NeglectStability` is marginally nicer than something like `[Neglect::Stability]`?

Reply:
>  If = [] worked, would that be preferred?

I'm inclined to say no. With the regular-types encoding, it's easy to restrict invocations of `transmute_{into,from}`




I think there may be additional shortcomings to encoding `Neglect` with const generics.

Reply:
>  If = [] worked, would that be preferred?

I'm inclined to say no. Another factor to consider is restricting invocations of `transmute_{from,into}` to combinations of options that do not jeopardize safety; e.g. note the `where` bound of `transmute_from`:
```rust
pub unsafe trait TransmuteFrom<Src: ?Sized, Scope, Neglect = ()>
where
    Neglect: TransmuteOptions,
{
    #[inline(always)]
    fn transmute_from(src: Src) -> Self
    where
        Src: Sized,
        Self: Sized,
        Neglect: SafeTransmuteOptions, /* ⬅ availability of this method depends on `Neglect` */
    {
        ...
    }

    #[inline(always)]
    unsafe fn unsafe_transmute_from(src: Src) -> Self
    where
        Src: Sized,
        Self: Sized,
        Neglect: TransmuteOptions,
    {
        ...
    }
}
```

This pattern is a little tricky to realize with const generics, since you cannot implement a trait (namely `SafeTransmuteOptions`) on *particular* const generic values.

Reply:
There's also a neat consideration with future possibilities.

Both a const `[]` representation and the current type tuple representations share a drawback: `Neglect` is, logically, a *set* of options, not an ordered sequence of options. There is no distinction between `[NeglectValidity,NeglectAlignment]` and `[NeglectAlignment,NeglectValidity]`. (I point this drawback out in the RFC [here](https://github.com/rust-lang/rfcs/blob/bb9b988c10a60d714ad52c9804aad36961960736/text/0000-safer-transmute.md#stability-of-unsafe-transmutations).) The ideal representation would use ad-hoc enum/union types, which Rust currently lacks.

If Rust ever gained such types, implementations of `TransmuteOptions` and `SafeTransmuteOptions` could be added for those types that would co-exist alongside the impls of those traits for tuples.

This would not be possible with a const generics representation of `Neglect`. Declaring `TransmuteFrom` to have the parameter `const NEGLECT: &'static [Option]` would commit us, forever, to encoding options as arrays.

Reply:
The "natural" representation for a type-level set like this is, arguably, a trait bound. It ignores order by definition, and trivially lets other parts of the program check for subsets or supersets, including lattice-style sets.

For example, the `Fn` traits are a simple example of this. There is no `FnOptions` trait implemented for tuples of marker types, there are just multiple traits with bounds describing the relationships between them.

The question here is thus whether anyone actually needs to be polymorphic over transmute options, or whether `TransmuteFrom` should simply be split up into separate traits without a `Neglect` parameter.

Reply:
> The question here is thus whether anyone actually needs to be polymorphic over transmute options, or whether TransmuteFrom should simply be split up into separate traits without a Neglect parameter.

It's not only a question of being polymorphic over neglected static checks. Since *combinations* of static checks may be neglected, it's a choice between a combinatorial explosion of trait impls (namely, an impl of `TransmuteOptions` for each combination of options), and a combinatorial explosion of transmutation traits (e.g., `TransmuteFrom`, `TransmuteFromNeglectValidity`,  `TransmuteFromNeglectAlignment`,  `TransmuteFromNeglectAlignmentValidity`, etc...).

My personal feeling is that new traits are much "weightier" than new impls. For instance, trait declarations dominate rustdoc in a way trait implementations do not. Most consumers probably will not want to neglect *any* static checks, but rustdoc does not have a mechanism to highlight a particular trait over others. In contrast, this RFC's `TransmuteFrom` default of `Neglect = ()` automatically selects the right set of options for most use cases. One can totally ignore the existence of that type parameter until they need to reach for it.

Reply:
Arguably the "weight" of a trait is a feature, not a bug- the options all still exist and must be considered regardless of how they are represented, so sweeping them under the rug by using a "lighter" language feature seems more like sleight-of-hand than an actual improvement to the API.

The existence of a combinatorial explosion is also somewhat concerning, IMO. If we don't need polymorphism over options, do we really need all the possible combinations as well?